src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/WorkspaceProjectContextHost.WorkspaceProjectContextHostInstance.cs

```diff
@@ -181,80 +194,77 @@ public async Task<T> OpenContextForWriteAsync<T>(Func<IWorkspaceProjectContextAc
-                    // version we are up to. Without this, operation progress might not complete.
-                    UpdateProgressRegistration();
-                    return;
-                }
+                    return ApplyInBatchAsync();
 
-                var state = new ContextState(isActiveEditorContext, isActiveConfiguration);
+                    async Task ApplyInBatchAsync()
+                    {
+                        IWorkspaceProjectContext context = _contextAccessor.Context;
 
-                _lastContextState = state;
+                        context.StartBatch();
 
-                IWorkspaceProjectContext context = _contextAccessor.Context;
+                        try
+                        {
+                            applyFunc(contextState, cancellationToken);
+                        }
+                        finally
+                        {
+                            await context.EndBatchAsync();
 
-                context.StartBatch();
+                            UpdateProgressRegistration();
+                        }
+                    }
 
-                try
-                {
-                    switch (handlerType)
+                    (ContextState ContextState, bool StateChanged) GetContextState(StrongBox<ContextState?> lastContextState)
                     {
-                        case WorkspaceContextHandlerType.Evaluation:
-                            _applyChangesToWorkspaceContext!.Value.ApplyProjectEvaluation(change.Subscription, state, cancellationToken);
-                            break;
-
-                        case WorkspaceContextHandlerType.ProjectBuild:
-                            Assumes.NotNull(change.CommandLineArgumentsSnapshot);
-                            _applyChangesToWorkspaceContext!.Value.ApplyProjectBuild(change.Subscription, change.CommandLineArgumentsSnapshot, state, cancellationToken);
-                            break;
-
-                        case WorkspaceContextHandlerType.SourceItems:
-                            _applyChangesToWorkspaceContext!.Value.ApplySourceItems(change.Subscription, state, cancellationToken);
-                            break;
+                        bool isActiveConfiguration = activeConfiguredProject == _project;
+                        bool isActiveEditorContext = _activeWorkspaceProjectContextTracker.IsActiveEditorContext(_contextAccessor.ContextId);
+                        
+                        ContextState newState = new(isActiveEditorContext, isActiveConfiguration);
+
+                        bool stateChanged = false;
+
+                        if (lastContextState.Value is null ||
+                            lastContextState.Value.Value.IsActiveConfiguration != newState.IsActiveConfiguration ||
+                            lastContextState.Value.Value.IsActiveEditorContext != newState.IsActiveEditorContext)
```

Comment:
The full type here is `StrongBox<ContextState?>`, where `ContextState` is a `struct`.

- The first `.Value` gets the value within the `StrongBox<T>`. We need that type in order to have ref semantics to a value within an async function.
- The second `.Value` is dereferencing the `Nullable<ContextState>` within the strong box.

Reply:
Ah. I see. So it would be equivalent to:
```C#
if(lastContextState.Value?.IsActiveConfiguration != newState.IsActiveConfiguration ||
   lastContextState.Value?.IsActiveEditorContext != newState.IsActiveEditorContext)
```

Reply:
Yes. That code doubles the null check though.